### PR TITLE
feat: remove `Js.Array2`, use labeled arguments for `Js.Array`

### PR DESCRIPTION
**Pull request by anmonteiro**
part of the unification proposed in https://github.com/melange-re/melange/issues/731

this needs the https://github.com/reasonml/reason-react/pull/821 in reason-react to compile the template projects.

### CODE REVIEW

File: jscomp/runtime/js_array.ml [hunk 1]
```diff
@@ -25,216 +25,193 @@
...
-[@@mel.send.pipe: 'a t as 'this]
+(** Mutating functions *)
+
+external copyWithin : 'a t -> to_:int -> 'a t = "copyWithin" [@@mel.send]
 (* ES2015 *)
 
-external copyWithinFrom : to_:int -> from:int -> 'this = "copyWithin"
-[@@mel.send.pipe: 'a t as 'this]
+external copyWithinFrom : 'a t -> to_:int -> from:int -> 'a t = "copyWithin"
+[@@mel.send]
```
Comment (jchavarri) [inline jscomp/runtime/js_array.ml:53]: I wonder if we apply this approach across the board for melange bindings, maybe we can get rid of `send.pipe` eventually? I recall Bob mentioning implementing it added a lot of complexity (plus, from user point of view having two ways to do the same thing is confusing).

File: jscomp/runtime/js_array.ml [hunk 2]
```diff
@@ -25,216 +25,193 @@
...
+external indexOf : 'a t -> value:'a -> int = "indexOf" [@@mel.send]
 
-external lastIndexOfFrom : 'a -> from:int -> int = "lastIndexOf"
-[@@mel.send.pipe: 'a t as 'this]
+external indexOfFrom : 'a t -> value:'a -> from:int -> int = "indexOf"
+[@@mel.send]
 
-external lastIndexOf_start : 'a -> int = "lastIndexOf"
-[@@mel.send.pipe: 'a t as 'this] [@@deprecated "Please use `lastIndexOf"]
+external join : 'a t -> sep:string -> string = "join" [@@mel.send]
```
Comment (jchavarri) [inline jscomp/runtime/js_array.ml:124]: Is it possible to have `sep` as optional? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join#separator

File: jscomp/runtime/js_array.ml [hunk 3]
```diff
@@ -25,216 +25,193 @@
...
-external slice_start : int -> 'this = "slice"
-[@@mel.send.pipe: 'a t as 'this] [@@deprecated "Please use `sliceFrom`"]
-
-external toString : string = "toString" [@@mel.send.pipe: 'a t as 'this]
-
-external toLocaleString : string = "toLocaleString"
-[@@mel.send.pipe: 'a t as 'this]
+external slice : 'a t -> start:int -> end_:int -> 'a t = "slice" [@@mel.send]
+external copy : 'a t -> 'a t = "slice" [@@mel.send]
+external sliceFrom : 'a t -> pos:int -> 'a t = "slice" [@@mel.send]
```
Comment (jchavarri) [inline jscomp/runtime/js_array.ml:132]: If we used `send.pipe` we could consolidate `slice` and `sliceFrom` under a single binding:

```ml
external slice : ?start:int -> ?end_:int -> 'a t = "slice" [@@mel.send.pipe: 'a t]
```

With `send` it requires unit at the end, but also works:

```ml
external slice : 'a t -> ?start:int -> ?end_:int -> unit -> 'a t = "slice" [@@mel.send]
```

At the very least, if we want to keep both `slice` and `sliceFrom`, I'd use `start` in `sliceFrom`, for consistency with `slice` and the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice#start).

File: jscomp/runtime/js_array.ml [hunk 4]
```diff
@@ -25,216 +25,193 @@
...
 (* ES2015 *)
 
-external copyWithinFrom : to_:int -> from:int -> 'this = "copyWithin"
-[@@mel.send.pipe: 'a t as 'this]
+external copyWithinFrom : 'a t -> to_:int -> from:int -> 'a t = "copyWithin"
+[@@mel.send]
 (* ES2015 *)
 
-external copyWithinFromRange : to_:int -> start:int -> end_:int -> 'this
+external copyWithinFromRange : 'a t -> to_:int -> start:int -> end_:int -> 'a t
```
Comment (jchavarri) [inline jscomp/runtime/js_array.ml:56]: Same ideas about optional labelled exposed in `slice` apply here.

File: jscomp/runtime/js_array.ml [hunk 5]
```diff
@@ -25,216 +25,193 @@
...
+[@@mel.send]
 (* ES2015 *)
 
-external fillInPlace : 'a -> 'this = "fill" [@@mel.send.pipe: 'a t as 'this]
+external fillInPlace : 'a t -> value:'a -> 'a t = "fill" [@@mel.send]
 (* ES2015 *)
 
-external fillFromInPlace : 'a -> from:int -> 'this = "fill"
-[@@mel.send.pipe: 'a t as 'this]
+external fillFromInPlace : 'a t -> value:'a -> from:int -> 'a t = "fill"
```
Comment (jchavarri) [inline jscomp/runtime/js_array.ml:64]: Same ideas about optional labelled exposed in `slice` apply here.

File: jscomp/runtime/js_array.ml [hunk 6]
```diff
@@ -25,216 +25,193 @@
...
 
-(* TODO: Not available in Node V4  *)
-external includes : 'a -> bool = "includes"
-[@@mel.send.pipe: 'a t as 'this]
-(** ES2016 *)
+external unshiftMany : 'a t -> values:'a array -> int = "unshift"
+[@@mel.send] [@@mel.splice]
 
-external indexOf : 'a -> int = "indexOf" [@@mel.send.pipe: 'a t as 'this]
+(** Accessor functions *)
```
Comment (jchavarri) [inline jscomp/runtime/js_array.ml:108]: Not sure `concat` is an accessor function.

**Review by jchavarri — COMMENTED** [review]
Looks great! I wonder if there's a chance to simplify the bindings by removing "dual" versions of the functions that can take optional args.

File: jscomp/runtime/js_array.ml (hunk 1 above)
Comment (anmonteiro) [reply jscomp/runtime/js_array.ml:53]: This probably warrants another discussion.

Here's the [PR](https://github.com/rescript-lang/rescript-compiler/pull/5146/files?w=1) where `send.pipe` got removed from ReScript. I don't think it's too much code to maintain, personally.

File: jscomp/runtime/js_array.ml (hunk 3 above)
Comment (anmonteiro) [reply jscomp/runtime/js_array.ml:132]: This is a good point, I like the suggestion, and I don't think adding an extra unit is too invasive (they're not compiled to extra arguments in JS).

The only question to solve is whether there's a difference between `['foo', 'bar'].join()` and `['foo','bar'].join(undefined)`

Comment (anmonteiro) [reply jscomp/runtime/js_array.ml:132]: Looks like for `join` there's no difference, will check `slice` etc.

**Pull request merged**